### DISCUSSION TITLE
How do I correctly get the socket instance from the socket.io server based on the socketID?

### DISCUSSION THREAD

**Discussion by JanYork** [socketio/socket.io#4982] [Q&A]
This is my dependency information, I am using socket in Nest.
```js
  "dependencies": {
    "@liaoliaots/nestjs-redis": "^9.0.5",
    "@nestjs/common": "^10.0.0",
    "@nestjs/core": "^10.0.0",
    "@nestjs/platform-express": "^10.0.0",
    "@nestjs/platform-socket.io": "^10.3.3",
    "@nestjs/websockets": "^10.3.3",
    "@types/deep-diff": "^1.0.5",
    "deep-diff": "^1.0.2",
    "ioredis": "^5.3.2",
    "lodash": "^4.17.21",
    "reflect-metadata": "^0.2.0",
    "rxjs": "^7.8.1",
    "socket.io": "^4.7.4"
  },
  "devDependencies": {
    "@nestjs/cli": "^10.0.0",
    "@nestjs/schematics": "^10.0.0",
    "@nestjs/testing": "^10.0.0",
    "@types/express": "^4.17.17",
    "@types/jest": "^29.5.2",
    "@types/node": "^20.3.1",
    "@types/supertest": "^6.0.0",
    "@typescript-eslint/eslint-plugin": "^6.0.0",
    "@typescript-eslint/parser": "^6.0.0",
    "eslint": "^8.42.0",
    "eslint-config-prettier": "^9.0.0",
    "eslint-plugin-prettier": "^5.0.0",
    "jest": "^29.5.0",
    "prettier": "^3.0.0",
    "source-map-support": "^0.5.21",
    "supertest": "^6.3.3",
    "ts-jest": "^29.1.0",
    "ts-loader": "^9.4.3",
    "ts-node": "^10.9.1",
    "tsconfig-paths": "^4.2.0",
    "typescript": "^5.1.3"
  },
```

I want to get the socket instance with the corresponding ID in the socket.io server through the socketID. The method is currently as follows: `const socket = first(await server.server.in(socketId).fetchSockets());`

This can be obtained and is normal until I need to clean up all rooms bound to a certain socket.
```js
socket.rooms.forEach((room) => {
      socket.leave(room);
});
```
When I removed all rooms and added new ones, I found that I could not obtain the socket instance through the above method. Why?

I tried to compare all the data of the socket.io server object and found that the rooms in the adapter have changed, as follows:
```js
// previous data
adapter: Adapter {
    // ......
    rooms: Map(1) { 'coBmXsIICq9iUTxAAAAD' => [Set] },
}

// data after
adapter: Adapter {
    // ......
    rooms: Map(1) { '4-all' => [Set] },
}
```

So I came to a result: `await server.server.in(socketId).fetchSockets()` is associated to the instance based on the k-v structure in rooms, but this is the easiest way to get the instance. Why does the author of socket.io Why not provide a more direct method without side effects? Or maybe I didn't find it?

I tried to find other methods, but I didn't find that the container that actually stores the Socket instance associated with the socketID is the sockets in the socket.io server. This is a Map, the socketID is the key, and the Socket instance is the value, but I found that socket.io The developer has disabled the get method of this Map, and it seems that they don't want me to get the socket instance from it.

But I tried to bypass this restriction, `const socket = (this.server?.sockets as any).get(id);` This way I can successfully get the data I want, but I think this is not the best way, because the author deliberately limited the get() of this Map.

In typescript, I saw a field prompt. There is also such a field, `this.server.sockets.sockets.get(id);`, this.server is the server of socket.io.
But `this.server.sockets` is a map, why do we need to add another socket to this map object? Moreover, this socket is also a Map, and its get() is publicly available. Unfortunately, this attribute is of no use. ``this.server.sockets.sockets`` is undefined data. Why? Wrong typescript type? Or is it intentional by the author of socket.io?

Well, I haven't found the optimal method yet. Can anyone give me an answer to this question?
1. How do I correctly find the socket instance by socketID without side effects (other associated effects).
2. The official documentation recommends using `const sockets = await io.in("room1").fetchSockets();` to obtain the instance, but when I remove all rooms of the socket, this method is not available. Why is this? Design this method?
3. I tried `const sockets = await io.of("/admin").in("room1").fetchSockets();` from the official documentation, but the compiler told me that of() uses undefined method, but typescript tells me that it exists, including the documentation.
4. Why should the get() of `this.server.sockets;` be disabled? And why `.sockets` is an undefined parameter, but it exists in the typescript type?

I'd like to get these answers, can anyone answer me? Thanks.

**Comment by darrachequesne** [comment]
Hi!

> How do I correctly find the socket instance by socketID without side effects (other associated effects).

`await server.in(socketId).fetchSockets()` is indeed the way to go.

> The official documentation recommends using const sockets = await io.in("room1").fetchSockets(); to obtain the instance, but when I remove all rooms of the socket, this method is not available. Why is this? Design this method?

You must not leave the default room (`coBmXsIICq9iUTxAAAAD` in your example).

```js
socket.rooms.forEach((room) => {
  if (room !== socket.id) {
    socket.leave(room);
  }
});
```

Depending on your use case, you may also use the `socketsLeave()` method:

```js
server.in(socketId).socketsLeave("some-room");
```

> I tried const sockets = await io.of("/admin").in("room1").fetchSockets(); from the official documentation, but the compiler told me that of() uses undefined method, but typescript tells me that it exists, including the documentation.

> Why should the get() of `this.server.sockets;` be disabled? And why `.sockets` is an undefined parameter, but it exists in the typescript type?


That should not happen. Could you please provide the full code reproducing the issue?

**Reply by JanYork** [reply]
<img width="1130" alt="image" src="https://github.com/socketio/socket.io/assets/88621545/903b1685-6bb7-4984-9561-f0554600d542">

`this.server` is an instance of socket.io, which I use in my Nest project.

`this.server.sockets.get()`, this method cannot be found, but `this.server.sockets` is a Map, so I do this (`(this.server?.sockets as any).get(id) `) to get it because I can't use its get() method.

Also, I can get a typescript prompt in `this.server.sockets`: there is a sockets attribute, so the code type of `this.server.sockets.sockets.get(id);` is correct, but `this .server.sockets.sockets` is an undefined property.

All codes are as follows:

```js
import {
  OnGatewayConnection,
  OnGatewayDisconnect,
  WebSocketGateway,
  WebSocketServer,
} from '@nestjs/websockets';
import { Server, Socket } from 'socket.io';
import { isUndefined } from '@nestjs/common/utils/shared.utils';

@WebSocketGateway({
  cors: {
    origin: '*',
  },
  namespace: '/socket',
  transports: ['websocket'],
})
export class SocketServer implements OnGatewayConnection, OnGatewayDisconnect {
  constructor() {}

  @WebSocketServer()
  public readonly server: Server;

  handleConnection = async (client: Socket): Promise<void> => {
    console.log('client connected', client.id);
  };

  handleDisconnect(client: any): any {
    console.log('client disconnected', client.id);
  }

  /**
   * remove all rooms
   *
   * @param id socket id
   */
  public async removeAllRooms(id: string): Promise<void> {
    const socket = await this.findSocketById(id);

    if (isUndefined(socket)) {
      return;
    }

    // remove all rooms
    socket.rooms.forEach((room) => {
      socket.leave(room);
    });
    console.log('all rooms are removed');
  }

  /**
   * find socket by id
   *
   * @param id socket id
   */
  public async findSocketById(id: string): Promise<Socket | never> {
    this.server.sockets.sockets.get(id);

    // const socket = first(await this.server.in(String(socketId)).fetchSockets());    The first function comes from Lodash.js.

    /**
     * There is no problem in getting it this way,
     * but the author of socket.io hides its get method and deliberately prevents developers from getting the socket instance from it.
     */
    const socket = (this.server?.sockets as any).get(id);

    if (isUndefined(socket)) {
      console.log('socket is not found');
      return null;
    }

    return socket as unknown as Socket;
  }

  /**
   * join room
   *
   * @param id socket id
   * @param room room name
   */
  public async joinRoom(id: string, room: string): Promise<boolean> {
    const socket = await this.findSocketById(id);
    if (isUndefined(socket)) {
      return false;
    }

    socket.join(room);
    return true;
  }
}
```

**Reply by JanYork** [reply]
`this.server.sockets.sockets.get(id);` In this code, `this.server.sockets.sockets` is an undefined property, and its type is a `Map<string, Socket<DefaultEventsMap, DefaultEventsMap, DefaultEventsMap, any>>`. When I try to run` this.server.sockets.sockets.get(id);`IDE prompts me: The get method cannot be called on an undefined property.

`this.server.sockets.sockets.get(id);` error message:
```js
[Nest] 33713  - 03/23/2024, 4:33:05 PM   ERROR [ExceptionsHandler] Cannot read properties of undefined (reading 'get')
TypeError: Cannot read properties of undefined (reading 'get')
    at SocketServer.findSocketById (/Users/mumu/Desktop/nest-socket/src/socket.server.ts:56:33)
    at AppController.getSocket (/Users/mumu/Desktop/nest-socket/src/app.controller.ts:19:44)
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-execution-context.js:38:29
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-execution-context.js:46:28
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-proxy.js:9:17
```

When I use `const socket = (this.server?.sockets as any).get(id);` to get the socket instance, I can get it, as shown in the figure:
<img width="1223" alt="image" src="https://github.com/socketio/socket.io/assets/88621545/9fbbb80f-fcbc-4cb8-a0f1-5930b3e6fc1d">

But the type of `this.server.sockets` is `Namespace<DefaultEventsMap, DefaultEventsMap, DefaultEventsMap, any>`, and the type of `this.server.sockets.sockets` is `Map<string, Socket<DefaultEventsMap, DefaultEventsMap, DefaultEventsMap, any>` >>`, why can I get the socket instance using `this.server.sockets` but `this.server.sockets.sockets` is an undefined property? From a type perspective, `this.server.sockets.sockets` is the container that stores socket instances.

**Comment by JanYork** [comment]
`this.server` is a `socket.io` instance，I simply used its function and didn't do anything else.

### one

```js
    await this.server.of('/admin').fetchSockets();
    console.log('server of：', this.server.of); // of is undefined
```
error message:
```js
TypeError: this.server.of is not a function
    at SocketServer.findSocketById (/Users/mumu/Desktop/nest-socket/src/socket.server.ts:57:23)
    at AppController.getSocket (/Users/mumu/Desktop/nest-socket/src/app.controller.ts:19:44)
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-execution-context.js:38:29
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-execution-context.js:46:28
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-proxy.js:9:17

```

### two

```js
 await this.server.of('/admin').in('room').fetchSockets();
```

error message:
```js
[Nest] 34525  - 03/23/2024, 4:47:48 PM   ERROR [ExceptionsHandler] this.server.of is not a function
TypeError: this.server.of is not a function
    at SocketServer.findSocketById (/Users/mumu/Desktop/nest-socket/src/socket.server.ts:57:23)
    at AppController.getSocket (/Users/mumu/Desktop/nest-socket/src/app.controller.ts:19:44)
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-execution-context.js:38:29
    at processTicksAndRejections (node:internal/process/task_queues:95:5)
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-execution-context.js:46:28
    at /Users/mumu/Desktop/nest-socket/node_modules/.pnpm/@nestjs+core@10.3.3_@nestjs+common@10.3.3_@nestjs+platform-express@10.3.3_@nestjs+websockets@_h3leh7b3mhe2cmhloksxb646bq/node_modules/@nestjs/core/router/router-proxy.js:9:17
```

**Reply by faltawy** [reply]
same here, have u managed to solve it ?

**Comment by Shakib448** [comment]
@JanYork 
You can like this way which is working in my environment

```
  private nsp: Namespace;
  @WebSocketServer()
  public readonly io: Namespace;
  
  
    afterInit() {
    this.nsp = this.io.server.of(your namespace);
  }
  
  public handleRoom() {
 this.nsp.emit("event", "working") 
 }

  
  ```

**Comment by andumy** [comment]
@JanYork 
I'm facing the same issue. My first assumption is that somewhere internally when the sockets map is assigned it goes to `server.sockets` instead of `server.sockets.sockets`. This would explain both why `server.sockets.sockets` is undefined as well as `server.sockets` being a map instead of a nampespace. 

I will investigate and raise a PR with a fix.

Meanwhile casting `server.sockets` as any seems to work as a workaround 👍
